tests/client/handlers/vaults.test.ts

```diff
@@ -1341,18 +1342,195 @@ describe('vaultsSecretsMkdir', () => {
+    const vaultIdEncoded = vaultsUtils.encodeVaultId(vaultId);
+    const dirPath = 'secret-first';
+    await vaultManager.withVaults([vaultId], async (vault) => {
+      await vault.writeF(async (efs) => {
+        await efs.writeFile(dirPath, dirPath);
+      });
+    });
+    // Attempt to make directory
+    const response = await rpcClient.methods.vaultsSecretsMkdir();
+    const writer = response.writable.getWriter();
+    await writer.write({ nameOrId: vaultIdEncoded, dirName: dirPath });
+    await writer.close();
+    // Check if the operation concluded as expected
+    for await (const data of response.readable) {
+      expect(data.success).toBeFalsy();
+      expect(data.error?.substring(0, 6)).toStrictEqual('EEXIST');
+    }
+    await vaultManager.withVaults([vaultId], async (vault) => {
+      await vault.readF(async (efs) => {
+        const stat = await efs.stat(dirPath);
+        expect(stat.isFile).toBeTruthy();
+      });
+    });
+  });
+  // test('makes directories in all vaults in a single log message', async () => {
+  //   const vaultName1 = 'test-vault1';
+  //   const vaultName2 = 'test-vault2';
+  //   const vaultId1 = await vaultManager.createVault(vaultName1);
+  //   const vaultId2 = await vaultManager.createVault(vaultName2);
+  //   const vaultIdEncoded1 = vaultsUtils.encodeVaultId(vaultId1);
+  //   const vaultIdEncoded2 = vaultsUtils.encodeVaultId(vaultId2);
+  //   const dirPath1 = 'dir-1';
+  //   const dirPath2 = 'dir-2';
+  //   const dirPath3 = 'dir-3';
+  //   let logLength1 = 0;
+  //   let logLength2 = 0;
+  //   await vaultManager.withVaults(
+  //     [vaultId1, vaultId2],
+  //     async (vault1, vault2) => {
+  //       const log1 = await vault1.log();
+  //       const log2 = await vault2.log();
+  //       logLength1 = log1.length;
+  //       logLength2 = log2.length;
+  //     },
+  //   );
+  //   // Attempt to make directories
+  //   const response = await rpcClient.methods.vaultsSecretsMkdir();
+  //   const writer = response.writable.getWriter();
+  //   await writer.write({
+  //     nameOrId: vaultIdEncoded1,
+  //     dirName: dirPath1,
+  //     metadata: { options: { recursive: true } },
+  //   });
+  //   await writer.write({ nameOrId: vaultIdEncoded2, dirName: dirPath2 });
+  //   await writer.write({ nameOrId: vaultIdEncoded1, dirName: dirPath3 });
+  //   await writer.close();
+  //   // Check if the operation concluded as expected
+  //   for await (const data of response.readable) {
+  //     expect(data.success).toBeTruthy();
+  //     expect(data.error).toBeUndefined();
+  //   }
+  //   await vaultManager.withVaults(
+  //     [vaultId1, vaultId2],
+  //     async (vault1, vault2) => {
+  //       const log1 = await vault1.log();
+  //       const log2 = await vault2.log();
+  //       expect(log1.length).toEqual(logLength1 + 1);
+  //       expect(log2.length).toEqual(logLength2 + 1);
+  //     },
+  //   );
+  // });
```

Comment:
I would assume that creating a directory would be modifying the vaults in some way, and would create a commit and increase the length of the log. However, I have noticed that the log length isn't changed, yet the directories still exist, weirdly enough.

I have looked at the code and confirmed that a commit is created after the callback function runs, but it clearly doesn't do anything in this case. If this shouldn't happen, I will keep on researching this issue. Otherwise, if this is intentional, I can just remove this test.

Reply:
That's a bug. Dig into the `writeF` code of `VaultInternal` and see what's happening.  We should see a commit if a directory is created. The fact that is is created without a new commit is doubly weird since that means it's in a dirty state as well and likely to be removed when it attempts a clean up.

Reply:
https://chatgpt.com/share/e/66ff4177-c57c-800d-a89b-50c9597d155d

Apparently git does not keep track of empty directories that it creates. They will remain invisible until a file is created within it, which will force the directory to also be recognised by git.

So, this is not an issue with `writeF`, but an issue with how git works itself. The solution is to make a placeholder file inside the directory, forcing git to stage that directory.

Reply:
This is possibly a problem that would allow a disconnect between the vault state and the clean vault history. We'll need to scope the problem out some more to see how it affects things. 

- Does the empty directory just exist within the vault? 
- What happens if we checkout a previous version? 
- Or checkout back to the latest version? 
- Does it get cloned when shared? 

Can we live with any of this? Or should we only allow directory creation when creating a new secret?

Reply:
> Does the empty directory just exist within the vault?

The empty directory exists within the vault file system. When running `pk secrets ls vault`, the newly created directory does show up in the output as a directory.

> What happens if we checkout a previous version?
> Or checkout back to the latest version?

If we checkout a previous version, the directory still exists in the vault fs. I tried rolling back to a previous version, and even back to the initial commit of the vault. The directory existed even there. This is most likely because the directory is outside git's version control tree, so it doesn't get deleted or modified when vault's version changes. Checking out to the latest version again yields a similar result of the empty directory being untouched.

> Does it get cloned when shared?

No, it doesn't. As the directory exists outside the git tree, it is not shared, and is not present when someone else is cloning the vault. I make a new local node, and tried to clone the vault with the empty directory.* It cloned everything but the empty directory. So, the empty directory only exists for the vault where it was initially created. 

This could pose issues where, if a vault's receiver manually made a similar directory in their vault, and the original directory got an update, then the incoming changes will conflict with the existing changes, causing all sorts of bad things.

\* The original vault which I use for testing purposes, `vault`, was unable to be shared via RPC cloning, even when both nodes were on the same device, and the vault only contains around 8 text files of a reasonable size. We may be able to use the `tar` compression to zip the vault fs for streaming, and converting the RPC over to, say, `RawHandler`.

Reply:
Yeah, I'm not a fan of this. There seems to be 3 options here. 

1. Just let there be a directory that's not part of the git tree. Not really good since its technically in a dirty vault state that we can't really manage.
2. Use a dummy special cased file in the directory when creating it. This would force the directory to be tracked by git but  we'd have to add a bunch of logic everywhere just to special case it and handle it. I really don't like it since it's prone to a buttload of edge cases and problems.
3. We scrap the `mkdir` command and only allow creating directories as part of creating secrets. So If I wanted a `newDir` i'd have to `secrets write vault:newdir/secret --recursive`, `secrets edit vault:newdir/secret --recursive` etc etc. This would be the best way to prevent the dirty vault state. `--recursive` would be needed because we generally don't implicitly create directories like this.

@CMCDragonkai Got any thoughts?

Reply:
Making an empty directory in git has no effect. This is part of the design of `git`.

It's an interesting conundrum here cause usually you'd create a directory and put some files in there AND then make the commit. But making a commit for just an empty directory is not possible.

So basically the issue is that empty directories basically do not exist as a concept in git, you may mutate the underlying vault and have these empty directories but they would not be committed to the log, nor would they be shared, if you have empty directories, it's as if they don't exist at all when you share the vault.

It's not a dirty vault state, it's not considered dirty at all. They are just invisible with respect to git and thus invisible with respect to git sharing/git log.

I think we keep the `mkdir`, but just that it doesn't by itself update the log. In that case you may need to skip the log commit system. It's an untracked invisible mutation basically. You keep it around, for as long as the the local vault exists, so I think solution 1. is the way to go.

I think you want to special case write a comment for this operation and indicate that empty are effectively non-existent in the vaults.

Reply:
> I think you want to special case write a comment for this operation and indicate that empty are effectively non-existent in the vaults.

I would say that the warning would be implemented on the CLI side, as the library itself shouldn't have this warning hard-coded. This must be up to the developers of the client.

As such, when I open a PR in Polykey CLI, I will have a reference to this comment, and add the warning/notice in that PR.

Reply:
> It's not a dirty vault state, it's not considered dirty at all. They are just invisible with respect to git and thus invisible with respect to git sharing/git log.

The `VaultsSecretsRemove` handler can also modify the vault state in a way such that we can be left with an empty directory. Do I need to warn users in that command too, that empty directories will not be a part of the source tree but will show up in the vault's file system when performing operations like listing contents, etc.?

Reply:
I think if our vaults operate like a regular git repo locally on a Unix system, then it's good enough. Basically people just need to realize that it's the same like a git repo.

Reply:
So, no need for a warning for other commands?

Reply:
I'm assuming that no warning is needed for other commands which might get the directories to a state as such.